src/nodes/poisson.jl

```diff
@@ -0,0 +1,29 @@
+export Poisson
+
+@node Poisson Stochastic [out, (l, aliases = [λ])]
+
+# ∑ [λ^k*log(k!)]/k! from k=0 to inf
+# Approximates the above sum for calculation of averageEnergy and differentialEntropy
+# @ref https://arxiv.org/pdf/1708.06394.pdf
+function __approximate_powersum(::Type{R}, l::T, j = 100) where {R, T}
+    if l == zero(T)
+        return zero(T)
+    elseif l > 110
+        # We give up on l>110, otherwise estimates are quite inaccurate
+        error("Cannot compute ∑ [λ^k*log(k!)]/k! for k > $l")
+    elseif l < 50 || R === BigFloat # asymptotic, does not work for large `l`
+        s = zero(R)
+        lk = one(R)
+        for k in 1:j
+            lk *= l
+            s += lk * loggamma(k + 1) / gamma(k + 1)
+        end
+        return s
+    else
+        # Try `BigFloat`
+        return convert(T, __approximate_powersum(BigFloat, l, 150))
+    end
+end
+
+@average_energy Poisson (q_out::Any, q_l::Any) =
+    mean(q_l) - mean(q_out) * mean(log, q_l) + exp(-mean(q_out)) * __approximate_powersum(Float64, mean(q_out))
```

Comment:
Derivation seems correct. I am not really up to speed of the approximation of `E_q(x)[ln(x!)]`, but could we also get away with the (perhaps simpler) Stirling's approximation? Furthermore I think we want to add some specialization for this average energy function, such that we do not always compute this approximation.

Reply:
Will it actually help?

Reply:
I think you still need to approximate this power sum even if `q_out` is a `PointMass`. We've spent quite some time yesterday to figure out how we can further simplify this expression. The computation is super unstable for large values. Distributions.jl has an interesting approximation here: https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/discrete/poisson.jl#L68, but I'm not sure how to incorporate it in our case (it should be possible, maybe I will spent more time on this). They do compute a slightly different quantity though. Any cool ideas are welcome!

Reply:
Ah, well, I understood you wrong. You are correct @bartvanerp if `Poisson` used for observations. The original approximation, however, assumes Poisson as a prior, not as a likelihood for your observations. So we need it anyway I believe. But it seems reasonable to add an extra dispatch method where `q_out` is a `PointMass`, it simplifies the computation indeed.